core/src/main/scala/kafka/admin/TopicCommand.scala

```diff
@@ -259,7 +263,8 @@ object TopicCommand extends Logging {
     override def alterTopic(opts: TopicCommandOptions): Unit = {
       val topic = new CommandTopicPartition(opts)
       val topics = getTopics(opts.topic, opts.excludeInternalTopics)
-      ensureTopicExists(topics, opts.topic)
+      ensureTopicExists(topics, opts.topic, !opts.ifExists)
```

Comment:
I don't think this is really enough to accomplish your goal here.  With this code, if the topic does not exist, we will get past this line but then fail on the next line in `describeTopics, with a `NoSuchTopicOrPartitionException`.

I think it would be better to simply catch the `NoSuchTopicOrPartitionException` and handle it appropriately like we did with topic creation.  This also avoids the redunant calls in `ensureTopicExists`.

Reply:
if the topic does not exist and , `ensureTopicExists()` will throw an `IllegalArgumentException` right? Let me look at this and the other two closely and get back

Reply:
Okay.. so I did think about the case where `--if-exists` is provided and thus `ensureTopicExists()` will move on without erroring... But I chose to let whatever exception that is thrown like NoSuchTopicOrPartitionException to propagate back to the caller/main() method..

I guess you are suggesting to go to the server first and then handle the exception..  I am fine changing it, it does seem cleaner.

Reply:
So this is an execution of `testAlterWhenTopicDoesntExistWithIfExists` altering a non-existing topic using `-if-exists` . As you can see, `topics` is empty.. The cleaner thing to do here seems to be just place any further calls into a `if (topics.nonEmpty) { .. }` block.. 

Is your intention here, avoiding the call to `getTopics()` and `ensureTopicExists()`?  That was the model I was referring to above. if you concern is `describeTopics` failing, then it does not seem to be happening.. (but this code is on thin-ice, agree. if someone changes behavior of describeTopics, then this will break, but so will the test I added.) 

![image](https://user-images.githubusercontent.com/1179324/83203309-038e0d80-a0fe-11ea-90a7-46abed714aa2.png)

Reply:
I was wondering about how the tests passed!  I guess I missed something important about `getTopics`-- the fact that it returns either one or zero topics.  (I forgot that zero was a possibility).

The problem with how the code works today is that it's really very wasteful.  It's literally listing every topic in the cluster, (which could be tens of thousands) just to see if the specific one you passed exists. There is also the potential for a TOCTOU issue here (for example, it existed but then was deleted right before you went to use it.)

So I would like to see this fixed to avoid doing the getTopics.  I guess think about it and if it seems reasonable to do it in the PR, let's go for it.  Otherwise I can create a follow-up JIRA.

We will need to fix stuff like this to scale the Kafka up.  People listing the full set of all topics is bad for scalability.

Reply:
Sounds good.  Thanks for digging into this!